### ISSUE TITLE
ol.style.Icon only works well when caching instances on the application level

### ISSUE THREAD

**Issue by michaelkrnac** [openlayers/openlayers#3137]
I add features to an Vectorlayer with Vector source.
With the Style function i display an SVG as icon.
When i display more than 33 features the icons start flickering.
I tried to debug this but there is no console error or something else.

![flikering](https://cloud.githubusercontent.com/assets/3944657/5761015/b9bc115e-9cd8-11e4-92e7-bbe74d87cb4d.gif)

```
  var styleFunction = function(feature, resolution) {

        var maxResolution = 500;
        if (feature.get('status') == "Valid")
            var textcolor = "#5cb85c";
        else if (feature.get('status') == "Draft")
            var textcolor = "#5bc0de";
        else if (feature.get('status') == "Invalid")
            var textcolor = "#d9534f";
        else if (feature.get('status') == "Soon Invalid")
            var textcolor = "#f0ad4e";
        else 
            var textcolor = "#FFF";

        var styleArray = [
            new ol.style.Style({
                image : new ol.style.Icon({
                    src : 'map' + feature.get('sysname')
                        + '/original.svg',
                    scale : maxResolution / resolution
                }),
                text: new ol.style.Text({
                    font: 'Calibri,sans-serif',
                    scale : 20000 / resolution,
                    offsetY: 500000 / resolution,
                    text: feature.get('name'),
                    fill: new ol.style.Fill({
                        color: textcolor
                    }),
                    stroke: new ol.style.Stroke({
                        color: "#FFF",
                        width: 5
                    })
                  })
            })
        ];
        return styleArray;
    }

    /**
     * MAP
     */
    var map = new ol.Map({
        target : document.getElementById('map'),
        layers : [
            new ol.layer.Tile({
                source : new ol.source.OSM()
            }), new ol.layer.Vector({
                title: 'Markers',
                source : new ol.source.Vector(),
                style: styleFunction
            })
        ],
        view : new ol.View({
            center : ol.proj.transform([
                1.80, 39.35
            ], 'EPSG:4326', 'EPSG:3857'),
            zoom : 2.9,
            minZoom : 2.5
        }),
        overlays : [
            popupoverlay
        ]
    });

    /**
     * SYTEMS
     */
    $.getJSON("map", function(data) {
        data.forEach(function(entry) {
            //console.log(entry);
            var lonlat = [
                parseFloat(entry.lon), parseFloat(entry.lat)
            ];
            var point = ol.proj.transform(lonlat, 'EPSG:4326', 'EPSG:3857');
            var feature = new ol.Feature({
                geometry : new ol.geom.Point(point),
                name : entry.name,
                sysname : entry.sysname,
                status : entry.status,
                id : entry.systemid
            });
            map.getLayers().a[1].getSource().addFeature(feature);
        });
    });
```

**Comment by ahocevar** [comment]
Try to cache your `ol.style.Icon` instead of creating a new one every time the style function is called. You can see how this is done in most of the style related vector examples.

Please reopen this issue with a comment if the above suggestion does not help.

**Comment by shaunakv1** [comment]
I am running into exact same issue. 33 unique image src's is the magic number and markers start flickering like christmas lights! I am doing the exact same thing as @michaelkrnac . Creating a `new ol.style.Icon` in the `new ol.style.Style`.  

@ahocevar  I am trying to refactor my code to cache `new ol.style.Icon` as you suggested, but problem is  `ol.style.Icon' only lets you set `size` , `opacity` and `src`  options when constructing it with `new`. Once it is constructed, the [API] (http://openlayers.org/en/v3.5.0/apidoc/ol.style.Icon.html) only allows setters on  `rotation` and `scale`, both of which are not enough to assign new images to different markers. Can you please show a small example of how to do cache?  Or point to which example is using this? 

Thanks!

**Comment by marcjansen** [comment]
We cannot diagnose your issue without code, @shaunakv1.

**Comment by shaunakv1** [comment]
I am sorry. I am doing exact same thing as the original question. But here's the code that is adding markers. 

```
function updateMarkers(markers){
				markersCopy = markers;
				infowindowEngine.setMarkers(markers);

				if(markerLayer) map.removeLayer(markerLayer);
				
				var vectorSource = new ol.source.GeoJSON(({ /** @type {olx.source.GeoJSONOptions} */
		            object: markers
		        }));

				 markerLayer = new ol.layer.Vector({
				    source: vectorSource,
				    style:function(feature, resolution){
				      var iconUrl = defaultMarkerIcon;
				      var iconWidth = defaultMarkerIconWidth;
				      var iconHeight= defaultMarkerIconHeight;

				      if(feature.getProperties().icon_info){
				        if(feature.getProperties().icon_info.image.url) 
				          iconUrl =  feature.getProperties().icon_info.image.url ;
				        
				        if(feature.getProperties().icon_info.width) 
				          iconWidth =  parseInt(feature.getProperties().icon_info.width);
				        
				        if(feature.getProperties().icon_info.height) 
				          iconHeight =  parseInt(feature.getProperties().icon_info.height);
				      }

				      return [new ol.style.Style({
				          image: new ol.style.Icon( /** @type {olx.style.IconOptions} */ ({
				              size:[iconWidth,iconHeight],
				              opacity: 1,
				              src: iconUrl
				          }))
				      })];
				    }
				});
				
				map.addLayer(markerLayer);
				map.getView().fitExtent(vectorSource.getExtent(), map.getSize());

				//check if user passed marker deep link parameter
				var m = p('marker');
				if (m !== 'null'){
					infowindowEngine.selectFeature(vectorSource.getFeatureById(m).getProperties(),vectorSource.getFeatureById(m).getProperties().geometry.getCoordinates());
				}
			}
```

**Comment by shaunakv1** [comment]
Okay so, I mis-understood what , @ahocevar meant by Caching the styles. After looking at this [example] (http://openlayers.org/en/v3.5.0/examples/kml-earthquakes.html)  I made slight modification to my code to implement style cache, and this works perfectly. No more flicker!   

If any one else runs into same problem, here's how I implemented the style caching
```
                                function(feature, resolution){	      
				      var style = styleCache[feature.getProperties().id];
				      if(style) return [style];
				      else{
				      		var iconUrl = defaultMarkerIcon;
				      		var iconWidth = defaultMarkerIconWidth;
				      		var iconHeight= defaultMarkerIconHeight;

				      		if(feature.getProperties().icon_info){
				      		  if(feature.getProperties().icon_info.image.url) 
				      		    iconUrl =  feature.getProperties().icon_info.image.url ;
				      		  
				      		  if(feature.getProperties().icon_info.width) 
				      		    iconWidth =  parseInt(feature.getProperties().icon_info.width);
				      		  
				      		  if(feature.getProperties().icon_info.height) 
				      		    iconHeight =  parseInt(feature.getProperties().icon_info.height);
				      		}
				      		styleCache[feature.getProperties().id] = new ol.style.Style({
				      		    image: new ol.style.Icon( /** @type {olx.style.IconOptions} */ ({
				      		        size:[iconWidth,iconHeight],
				      		        opacity: 1,
				      		        src: iconUrl
				      		    }))
				      		});
				      		return [ styleCache[feature.getProperties().id] ];
				      }
				    } // end of style function

```



How ever can't this be implemented inside the library, based on object signatures?  If nothing else, I think it would be nice if the documentation is updated to clearly show users how to do this and why it is needed.

**Comment by marcjansen** [comment]
Cool. So nice that you fixed it!

**Comment by marcjansen** [comment]
PRs improving the documentation are always welcome.

**Comment by ppkoonings** [comment]
I also have the same problem with flickering icons. I passed the 32 different icons, and then it starts.
At: http://www.welkomewijken.nl/index.php?kaart=2   , you can see the problem. Because I change the size of the icons, based on resolution, I am not able to use stylecaching. I tried OL v3.7.0 but problem remains... Somebody an idea what to do?

**Comment by ahocevar** [comment]
@ppkoonings Cache your icon style, and only call `setSize` on it in the style function.

**Comment by ppkoonings** [comment]
@ahocevar Thanks Andres, I was not able to use setSize (how?) but I fixed it with style.e.j = 1.2;

Here is the code:

var styleCache = {};
var resultOverlay = new ol.layer.Vector({
  title: "resultaat",
  source: resultsource,

   style: function(feature, resolution){

  var style = styleCache[ feature.get('gid') ];

  if(style)
  {
     if( feature.get("filter") == 0 )
     {
        return [  new ol.style.Style({
    stroke: new ol.style.Stroke({
      color: '#00FFFF',
      width: 3
    })
  })  ];
     }
     if( feature.get("filter") == 2 )
     {
       style.e.j = 1.2;
       console.log( style.e.j );

       return [style];
     }
     else
     {
        if( resolution < 6 )
        {
          style.e.j = 0.8;
        }
        else
        {
          style.e.j = 0.5;
        }

        return [style];
     }
  }
  else{

        styleCache[ feature.get('gid') ] = new ol.style.Style({
            image: new ol.style.Icon( /** @type {olx.style.IconOptions} */ ({
                 anchor: [16, 37],
                 scale: 0.5,
                 anchorXUnits: 'pixels',
                 anchorYUnits: 'pixels',
			     opacity: 0.90,
			     src: "icons/" + feature.get('kaart') + ".png"
            }))
        });

        return [ styleCache[  feature.get('gid') ] ];
  }

} // end of style function


});

**Comment by ahocevar** [comment]
You should not be accessing object properties like this. But you're right, we don't have a setSize() method. We should add one. Are you able to create a pull request? If not let me know, then I'll create one.

**Comment by ppkoonings** [comment]
Andreas, I never created a request, so I don't know how and where to do... I am not really familiair with forums and so..
Till now I never have used the object properties like this, but I do not know how to solve it in another way. Looking with console.log to the object I tried this, and it solved my problem.... No christmas tree anymore... Do you want to create the pull request?

**Comment by megaten** [comment]
Thanks for that thread. I've solve my problem with that sample, but there is an followup to solve that issue?

**Comment by RomainVernoux** [comment]
I confirm this is still happening in ol 3.11.2. Unfortunately I cannot use the cache workaround since we use clustering and icons are generated dynamically depending on the composition of the cluster.

Is there a way to fix the root issue instead of working around it?

**Comment by megaten** [comment]
To day, I run in 3.13 and the flickering is away without any change on my code, but I use the styleCache. ( more than 300/400 icons on the screen)

**Comment by eflowbeach** [comment]
I have the same issue as @RomainVernoux , my icons get generated on the fly via a php imaging service. My icons are wind barbs which have other meteorological data in them, so caching is not a solution. I was able to get this same functionality to work in OL2 with no flickering.

**Comment by ahocevar** [comment]
An icon can have an image and a size. You should be able to cache icons by these properties.

**Comment by RomainVernoux** [comment]
@eflowbeach I ended up caching the images anyway. What I did was to take into account all parameters that influence the image result (e.g. input data, size, colors...) and cache the icons by these properties. The flickering issue was solved but the cached data can grow quickly. The next step could be for me to implement a FIFO-like caching system to delete cached icons over time.

But I still believe we are talking about a workaround here and that the flickering is a bug that should be fixed inside OL3. Or maybe I am missing something?
